apollo-router/src/plugins/file_uploads/mod.rs

```diff
@@ -0,0 +1,644 @@
+
+        let request_body = hyper::Body::wrap_stream(operations_stream);
+        return Ok(router::Request::from((
+            http::Request::from_parts(request_parts, request_body),
+            req.context,
+        )));
+    }
+
+    Ok(req)
+}
+
+struct RouterLayerResult {
+    multipart: MultipartRequest,
+}
+
+async fn supergraph_layer(mut req: supergraph::Request) -> UploadResult<supergraph::Request> {
+    let service_layer_result = req
+        .context
+        .extensions()
+        .lock()
+        .remove::<RouterLayerResult>();
+
+    if let Some(RouterLayerResult { mut multipart }) = service_layer_result {
+        let map_field = multipart.map_field().await?;
+        let variables = &mut req.supergraph_request.body_mut().variables;
+        let mut files_order = IndexSet::new();
+        let mut map_per_variable: MapPerVariable = HashMap::new();
+        for (filename, paths) in map_field.into_iter() {
+            for path in paths.into_iter() {
+                let mut segments = path.split('.');
+                let first_segment = segments.next();
+                if first_segment != Some("variables") {
+                    if first_segment
+                        .and_then(|str| str.parse::<usize>().ok())
+                        .is_some()
+                    {
+                        return Err(FileUploadError::BatchRequestAreNotSupported);
+                    }
+                    return Err(FileUploadError::InvalidPathInsideMapField(path));
+                }
+                let variable_name = segments.next().ok_or_else(|| {
+                    FileUploadError::MissingVariableNameInsideMapField(path.clone())
+                })?;
+                let variable_path: Vec<String> = segments.map(str::to_owned).collect();
+
+                // patch variables to pass validation
+                let json_value = variables
+                    .get_mut(variable_name)
+                    .and_then(|root| try_path(root, &variable_path))
+                    .ok_or_else(|| FileUploadError::InputValueNotFound(path.clone()))?;
+                drop(core::mem::replace(
+                    json_value,
+                    serde_json_bytes::Value::String(
+                        format!("<Placeholder for file '{}'>", filename).into(),
+                    ),
+                ));
+
+                map_per_variable
+                    .entry(variable_name.to_owned())
+                    .or_insert_with(|| HashMap::new())
+                    .entry(filename.clone())
+                    .or_insert_with(|| Vec::new())
+                    .push(variable_path);
+            }
+            files_order.insert(filename);
+        }
+
+        req.context
+            .extensions()
+            .lock()
+            .insert(SupergraphLayerResult {
+                multipart,
+                map_per_variable,
+                files_order,
+            });
+    }
+    Ok(req)
+}
+
+fn try_path<'a>(
```

Comment:
It would be nice to add a comment to document what this function does

Reply:
https://github.com/apollographql/router/commit/367ea94546ff3f4db9c41d7e0f81e9884beddc10#diff-7067e2dd8c31cd5cd84b593d83c0002ce9a256d8d251abdf738bc83cc65e1bfeR250


we now have:

```rust
// Replaces value at path with the provided one.
// Returns the provided path if the path is not valid for the given object
fn replace_value_at_path<'a>(
    variables: &'a mut json_ext::Object,
    path: &'a [String],
    value: serde_json_bytes::Value,
) -> std::result::Result<(), &'a [String]> {
```

and 

```rust
// Removes value at path.
fn remove_value_at_path<'a>(variables: &'a mut json_ext::Object, path: &'a [String]) {
```

which both use Ivan's new `get_value_at_path` function